File: app/src/main/java/com/vladimir_x/easyenglishlearn/category_select/CategoryFragment.java
package com.vladimir_x.easyenglishlearn.category_select;

import androidx.lifecycle.ViewModelProvider;

import android.content.Context;

import androidx.databinding.DataBindingUtil;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.vladimir_x.easyenglishlearn.R;
import com.vladimir_x.easyenglishlearn.databinding.FragmentCategorySelectBinding;
import com.vladimir_x.easyenglishlearn.databinding.RvCategoryItemBinding;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import static com.vladimir_x.easyenglishlearn.Constants.DIALOG_REMOVE_CATEGORY;

public class CategoryFragment extends Fragment {

    private Callbacks mCallbacks;
    private CategoryViewModel mViewModel;
    private FragmentCategorySelectBinding mBinding;
    private CategoryAdapter mAdapter;

    public interface Callbacks {
        void onCategorySelected(String categoryName);

        void onCategoryEdit(String categoryName);
    }


    @NonNull
    public static Fragment newInstance() {
        return new CategoryFragment();
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        mCallbacks = (Callbacks) context;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container,
                             Bundle savedInstanceState) {
        mBinding = DataBindingUtil.inflate(inflater,
                R.layout.fragment_category_select,
                container,
                false);
        mBinding.rvCategorySelect.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (dy > 0 && mBinding.fabCategoryAdd.getVisibility() == View.VISIBLE) {
                    mBinding.fabCategoryAdd.hide();
                } else if (dy < 0 && mBinding.fabCategoryAdd.getVisibility() != View.VISIBLE) {
                    mBinding.fabCategoryAdd.show();
                }
            }
        });
        return mBinding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mViewModel = new ViewModelProvider(this).get(CategoryViewModel.class);
        mBinding.setViewModel(mViewModel);

        mBinding.rvCategorySelect.setLayoutManager(new LinearLayoutManager(getActivity()));
        mAdapter = new CategoryAdapter();
        mBinding.rvCategorySelect.setAdapter(mAdapter);
        subscribeToLiveData();
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mCallbacks = null;
    }

    private void subscribeToLiveData() {
        mViewModel.getCategoriesLiveData().observe(getViewLifecycleOwner(), mAdapter::setCategoryList);
        mViewModel.getEditCategoryLiveData().observe(getViewLifecycleOwner(), mCallbacks::onCategoryEdit);
        mViewModel.getRemoveDialogLiveData().observe(getViewLifecycleOwner(), this::showDialog);
        mViewModel.getOpenCategoryLiveData().observe(getViewLifecycleOwner(), mCallbacks::onCategorySelected);
        mViewModel.getMessageLiveData().observe(getViewLifecycleOwner(), this::showMessage);
    }

    private void showMessage(@StringRes int resId) {
        Toast.makeText(getActivity(), resId, Toast.LENGTH_SHORT).show();
    }

    private void showDialog(String categoryName) {
        DialogFragment dialogFragment = CategoryRemoveFragment.newInstance(categoryName);
        dialogFragment.show(requireActivity().getSupportFragmentManager(),
                DIALOG_REMOVE_CATEGORY);
    }

    private class CategoryAdapter extends RecyclerView.Adapter<CategoryHolder> {

        private List<String> mCategoryList = new ArrayList<>();


        @NonNull
        @Override
        public CategoryHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            LayoutInflater inflater = LayoutInflater.from(getActivity());
            RvCategoryItemBinding binding = DataBindingUtil.inflate(inflater,
                    R.layout.rv_category_item,
                    parent,
                    false);
            return new CategoryHolder(binding);
        }

        @Override
        public void onBindViewHolder(@NonNull CategoryHolder categoryHolder, int position) {
            categoryHolder.bind(mCategoryList.get(position));
        }

        @Override
        public int getItemCount() {
            return mCategoryList.size();
        }

        void setCategoryList(List<String> categoryList) {
            mCategoryList = categoryList;
            notifyDataSetChanged();
        }
    }

    private class CategoryHolder extends RecyclerView.ViewHolder {

        private RvCategoryItemBinding mBinding;


        CategoryHolder(RvCategoryItemBinding binding) {
            super(binding.getRoot());
            mBinding = binding;
            mBinding.setViewModel(mViewModel);
        }

        void bind(String categoryName) {
            mBinding.setCategoryName(categoryName);
            mBinding.executePendingBindings();
        }
    }
}

File: app/src/main/java/com/vladimir_x/easyenglishlearn/ui/category_select/CategoryFragment.kt
package com.vladimir_x.easyenglishlearn.ui.category_select

import android.content.Context
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.vladimir_x.easyenglishlearn.Constants
import com.vladimir_x.easyenglishlearn.R
import com.vladimir_x.easyenglishlearn.databinding.FragmentCategorySelectBinding
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class CategoryFragment : Fragment(R.layout.fragment_category_select) {
    private var callbacks: Callbacks? = null
    private var adapter: CategoryAdapter? = null
    private var _binding: FragmentCategorySelectBinding? = null
    private val binding get() = _binding!!
    private val viewModel: CategoryViewModel by viewModels()

    interface Callbacks {
        fun onCategorySelected(categoryName: String?)
        fun onCategoryEdit(categoryName: String?)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        callbacks = context as Callbacks
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        _binding = FragmentCategorySelectBinding.bind(view)
        binding.rvCategorySelect.layoutManager = LinearLayoutManager(activity)
        adapter = CategoryAdapter(
            { viewModel.onItemClick(it) },
            { viewModel.onEditClick(it) },
            { viewModel.onRemoveClick(it) }
        )
        binding.rvCategorySelect.adapter = adapter

        binding.rvCategorySelect.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                if (dy > 0 && binding.fabCategoryAdd.visibility == View.VISIBLE) {
                    binding.fabCategoryAdd.hide()
                } else if (dy < 0 && binding.fabCategoryAdd.visibility != View.VISIBLE) {
                    binding.fabCategoryAdd.show()
                }
            }
        })

        binding.fabCategoryAdd.setOnClickListener { viewModel.onFabClick() }
        subscribeObservers()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onDetach() {
        super.onDetach()
        callbacks = null
    }

    private fun subscribeObservers() {
        with(viewModel) {
            viewLifecycleOwner.lifecycleScope.launch {
                viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.RESUMED) {
                    launch {
                        categories.collect { categoryList ->
                            adapter?.setCategoryList(categoryList)
                        }
                    }
                    launch {
                        categoryState.collect { state ->
                            when (state) {
                                is CategorySelectState.OpenCategory -> {
                                    callbacks?.onCategorySelected(state.data)
                                }
                                is CategorySelectState.EditCategory -> {
                                    callbacks?.onCategoryEdit(state.data)
                                }
                                is CategorySelectState.RemoveCategory -> {
                                    showDialog(state.data)
                                }
                                else -> {}
                            }
                        }
                    }
                }
            }
        }
    }

    private fun showMessage(message: String) {
        Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
    }

    private fun showDialog(categoryName: String) {
        childFragmentManager.setFragmentResultListener(
            Constants.DIALOG_REMOVE_CATEGORY,
            viewLifecycleOwner
        ) { _, bundle ->
            when (bundle.getSerializable(Constants.RESULT_KEY) as DialogResult) {
                DialogResult.Yes -> dialogYesClicked(categoryName)
                else -> {}
            }
        }

        val dialogFragment = CategoryRemoveFragment.newInstance(categoryName)
        dialogFragment.show(
            childFragmentManager,
            Constants.DIALOG_REMOVE_CATEGORY
        )
    }

    private fun dialogYesClicked(categoryName: String) {
        viewModel.removeCategory(categoryName)
        showMessage(getString(R.string.category_removed, categoryName))
    }

    companion object {
        @JvmStatic
        fun newInstance(): Fragment {
            return CategoryFragment()
        }
    }
}